programs/showhostkey/showhostkey.8
'\" t
.\"     Title: IPSEC_SHOWHOSTKEY
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 10/06/2010
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "IPSEC_SHOWHOSTKEY" "8" "10/06/2010" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ipsec_showhostkey \- show host\'s authentication key
.SH "SYNOPSIS"
.HP \w'\fBipsec\fR\ 'u
\fBipsec\fR \fIshowhostkey\fR [\-\-ipseckey\ [\fIgateway\fR]] [\-\-left] [\-\-right] [\-\-dump] [\-\-verbose] [\-\-version] [\-\-list] [\-\-x509self] [\-\-x509req] [\-\-x509cert] [\-\-txt\ \fIgateway\fR] [\-\-dhclient] [\-\-file\ \fIsecretfile\fR] [\-\-keynum\ \fIcount\fR] [\-\-id\ \fIidentity\fR]
.SH "DESCRIPTION"
.PP
\fIShowhostkey\fR
outputs (on standard output) a public key suitable for this host, in the format specified, using the host key information stored in
/etc/ipsec\&.secrets\&. In general only the super\-user can run this command, since only he can read
\fIipsec\&.secrets\fR\&.
.PP
The
\fB\-\-txt\fR
option causes the output to be in opportunistic\-encryption DNS TXT record format, with the specified
\fIgateway\fR
value\&. If information about how the key was generated is available, that is provided as a DNS\-file comment\&. For example,
\fB\-\-txt 10\&.11\&.12\&.13\fR
might give (with the key data trimmed for clarity):
.sp
.if n \{\
.RS 4
.\}
.nf
  ; RSA 2048 bits   xy\&.example\&.com   Sat Apr 15 13:53:22 2000
      IN TXT  "X\-IPsec\-Server(10)=10\&.11\&.12\&.13 AQOF8tZ2\&.\&.\&.+buFuFn/"
.fi
.if n \{\
.RE
.\}
.PP
No name is supplied in the TXT record because there are too many possibilities, depending on how it will be used\&. If the text string is longer than 255 bytes, it is split up into multiple strings (matching the restrictions of the DNS TXT binary format)\&. If any split is needed, the first split will be at the start of the key: this increases the chances that later hand editing will work\&.
.PP
The
\fB\-\-version\fR
option causes the version of the binary to be emitted, and nothing else\&.
.PP
The
\fB\-\-verbose\fR
may be present one or more times\&. Each occurance increases the verbosity level\&.
.PP
The
\fB\-\-left\fR
and
\fB\-\-right\fR
options cause the output to be in
\fBipsec.conf\fR(5)
format, as a
\fBleftrsasigkey\fR
or
\fBrightrsasigkey\fR
parameter respectively\&. Again, generation information is included if available\&. For example,
\fB\-\-left\fR
might give (with the key data trimmed down for clarity):
.sp
.if n \{\
.RS 4
.\}
.nf
  # RSA 2048 bits   xy\&.example\&.com   Sat Apr 15 13:53:22 2000
  leftrsasigkey=0sAQOF8tZ2\&.\&.\&.+buFuFn/
.fi
.if n \{\
.RE
.\}
.PP
The
\fB\-\-dhclient\fR
option cause the output to be suitable for inclusion in
\fBdhclient.conf\fR(5)
as part of configuring WAVEsec\&. See <\m[blue]\fBhttp://www\&.wavesec\&.org\fR\m[]>\&.
.PP
If
\fB\-\-ipseckey\fR
is specified, the output format is the text form of a DNS IPSECKEY record (see RFC4025); the host name is the one included in the key information (or, if that is not available, the output of
\fBhostname\ \&\-\-fqdn\fR), with a
\fB\&.\fR
appended\&. The gateway information, if provided, is is included, otherwise, the gateway is assumed to be self, and to be of type FQDN\&. Generation information is included if available\&. For example (with the key data trimmed down for clarity):
.sp
.if n \{\
.RS 4
.\}
.nf
  ; RSA 2048 bits   xy\&.example\&.com   Sat Apr 15 13:53:22 2000
  xy\&.example\&.com\&.   IN   IPSECKEYKEY   floyd albert
.fi
.if n \{\
.RE
.\}
.sp
.if n \{\
.RS 4
.\}
.nf
  ; RSA 2048 bits   xy\&.example\&.com   Sat Apr 15 13:53:22 2000
  xy\&.example\&.com\&.   IN   KEY   0x4200 4 1 AQOF8tZ2\&.\&.\&.+buFuFn/
.fi
.if n \{\
.RE
.\}
.PP
Normally, the default key for this host (the one with no host identities specified for it) is the one extracted\&. The
\fB\-\-id\fR
option overrides this, causing extraction of the key labeled with the specified
\fIidentity\fR, if any\&. The specified
\fIidentity\fR
must
\fIexactly\fR
match the identity in the file; in particular, the comparison is case\-sensitive\&.
.PP
There may also be multiple keys with the same identity\&. All keys are numbered based upon their linear sequence in the file (including all include directives)
.PP
The
\fB\-\-file\fR
option overrides the default for where the key information should be found, and takes it from the specified
\fIsecretfile\fR\&.
.SH "DIAGNOSTICS"
.PP
A complaint about \(lqno pubkey line found\(rq indicates that the host has a key but it was generated with an old version of FreeS/WAN and does not contain the information that
\fIshowhostkey\fR
needs\&.
.SH "FILES"
.PP
/etc/ipsec\&.secrets
.SH "SEE ALSO"
.PP
ipsec\&.\fBsecrets\fR(5), ipsec\&.\fBconf\fR(5),
\fBipsec_rsasigkey\fR(8)
.SH "HISTORY"
.PP
Written for the Linux FreeS/WAN project <\m[blue]\fBhttp://www\&.freeswan\&.org\fR\m[]> by Henry Spencer\&.
.SH "BUGS"
.PP
Arguably, rather than just reporting the no\-IN\-KEY\-line\-found problem,
\fIshowhostkey\fR
should be smart enough to run the existing key through
\fIrsasigkey\fR
with the
\fB\-\-oldkey\fR
option, to generate a suitable output line\&.
.PP
The need to specify the gateway address (etc\&.) for
\fB\-\-txt\fR
is annoying, but there is no good way to determine it automatically\&.
.PP
There should be a way to specify the priority value for TXT records; currently it is hardwired to
10\&.
.PP
The
\fB\-\-id\fR
option assumes that the
\fIidentity\fR
appears on the same line as the
\fB:\ \&RSA\ \&{\fR
that begins the key proper\&.


CHANGES

Note that listings are in chronological order of release times, not in order
of version numbers, so you will find 2.5.x and 2.6.x releases intersperced.

See also docs/KNOWN_BUGS and http://bugs.openswan.org/

v2.6.30 (unreleased)
* Add geode-aes to the list of crypto modules to load [Paul]
* Add --listen/listen= option to limit listening to a single IP [Paul]
* sa_policy_bit_names was missing ModeConfig DNS and WINS bit names [Paul]
* SAREF: Add sareftrack=<no|yes|conntrack> connection option [Paul]
* Fix for: ASSERTION FAILED at connections.c:1579: c->kind == CK_TEMPLATE [Paul]
* Fix for "handling event EVENT_RETRANSMIT for <invalid>" [Paul]
* Fix for specifying protport=47 (GRE has no ports) [Paul]
* Bugtracker bugs fixed:
   # 252/619 more than 20 payloads in message; ignored [paul]
   # 690 ipsec lwdnsq --help not implemented
   # 860 Port --random for newhostkey [Paul]
   #1005 Incorrect message "R_U_THERE_ACK has unexpected sequence number" [Mike]
   #1040: Fix to compile without DEBUG [Paul]
   #1054: Startup warning: "ignored obsolete keyword (null)" [Michael Smith]
   #1112: Prototypes only, if function enabled in c-source with KLIPS or
          PFKEY [Henry N.]
   #1115: Fix various warnings u_char * vs. char * for sscanf,... [Henry N.]
   #1149: pluto uses empty NAT_OA as IDci, Server behind NAT and non natted
          Windows XP [Wolfgang Nothdurft]
   #1151: The ipsec module is not removed by 'ipsec setup stop' [Paul]

v2.6.29 (September 27, 2010)
! This release is made for CVE-2010-3302 and CVE-2010-3308
! Duplicate CVE's issues for these are CVE-2010-3752 and CVE-2010-3753
* XAUTH: Avoid buffer overflow in CISCO DNS info [dhr/paul]
         Avoid shell problems with single quotes CISCO DNS paramters [dhr/paul]
  http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2010-3302
* XAUTH: Avoid buffer overflow in CISCO BANNER [dhr/paul]
         Avoid shell problems with single quotes in CISCO paramters [dhr/paul]
  http://www.cve.mitre.org/cgi-bin/cvename.cgi?name=2010-3308
* NETKEY: Fix for spurious %hold netlink-acquires [Paul/dhr]
* KLIPS: Fix compiling on 2.6.18 based RHEL5 kernels [Paul]
* Various fixes based on automated source code review [dhr]
* SAREF: Updated for 2.6.35 kernel [Harald]
* KLIPS: Updated for 2.6.35 kernel [Harald]
* PACKAGING Use Epoch 1: for Debian/Ubuntu [Simon]
* MAST: fix iptables rule "leak" on rekey [Bart]
* MAST: use only the most recent iptables rule [Bart]
* pluto: restrict rekeymargin to be smaller than salifetime [Bart]
* MAST: ensure we don't end up with mtu=0 on mast0 [Bart/Paul]
* MAST: enforce outgoing tunnel policy [Bart]
* MAST: use addflow pfkey command to set policy on tunnel SAs [Bart]
* Added a new pfkey flag, POLICYONLY, to the ADDFLOW command [Bart]
* MAST: allow for setting of policy for inbound SAs [Bart]
* MAST: favour deleting an SA even if the pfkey op failed [Bart]
* HAVESTATSD: Log new phase2 messages as a result of a rekey [Paul]
* MAST: use iptables --comment to show the conn name [Bart]
* VNET: differentiate instantiation of road warriors and vnet [Paul]
* Log LEAK_DETECTIVE and HAVE_LIBNSS support on startup [Paul]
* [IKEv2] connections were broken since 2.6.25 [Avesh]
* MAST: new "ipsec policy" command replaces "ipsec eroute" [Bart]
* Fix SElinux warning in realsetup (bz628879) [Avesh]
* Support for SHA2_256 in IKEv2 (bz621790) [Avesh]
* IKEv2: Fix for using MD5 and PRF conversion function [Avesh]
* SAREF: Improved workaround for rp_filter [Bart]
* NSS: Increase minimum nss for rhbz#453577 [Paul]
  (this allows us to revert workaround in git 6c8ff2791d1)
* SAREF: Added /proc/net/ipsec/saref that shows kernel patch state [Bart]
* SAREF: Add SAref checks to ipsec verify [bart/paul]
* A bunch of Makefile.inc variables could not be overridden properly [Paul]
* Bugtracker bugs fixed:
   #1042 NF_INET_LOCAL_OUT vs NF_IP_LOCAL_OUT still broken [David]
   #1102: Ensure KLIPS still works on older kernels [Paul]
   #1113: crypt_dh.c: Fix "warning: multi-line comment" [Tuomo]

v2.6.28 (July 29, 2010)
* MAST: Fix SAref vmalloc() call that could cause a kernel panic [Bart]
* SAREF: rework exported interface to avoid might_sleep() during rcu lock [Bart]
* SAREF: Use the correct mask when updating nfmark in rcv path [Bart]
* MAST: Cleanup updown.mast iptables rule management [Bart]
* MAST: Rework mast init scripts to use conntrack [Bart]
* MAST: Remove iptables rules after SA is deleted [Bart]
* SAREF: Fix bug in stream-socket saref mode [Bart]
* SAREF: Ported to apply on Linux 2.6.34 [Harald]
* MAST: Fix for NAT-T mode (set ixs->outgoing_said in mast mode) [Bart]
* MAST: Fix module unload with mast [Bart]
* Allow rightsubnet=vnet:%priv rightprotoport=17/%any w/o right=%any [Paul]
* SAREF: Log SAref and SAbind capabilities on pluto startup [Paul]
* Log tunnel down with HAVE_STATSD as "down", not "unknown" [Paul]
* Changes to _realsetup.in for making the init script LSB compliant [Avesh]
  rhbz #594767
* _startnetkey update for DNS and NetworkManager [Avesh]
* SAREF: fix bug in stream-socket saref mode [Bart]
* Added TCP header flags to ipsec_print_ip() output [Bart]
* KLIPS: Don't fail non-existant header_ops (breaks ppp) [David]
* KLIPS: ip_select_ident hashing fix in AH xmit path [Kirill Berezin]
* HAVE_STATSD: Fix phase1/phase2 logging through HAVE_STATSD interface [David]
* HAVE_STATSD: log output fix when two connections share a phase1 [David]
* HAVE_STATSD: Slightly clarified and changed log messages [Paul]
* SAREF: Clarified defines and fixed nfmark printing in HAVE_STATSD [dhr/Paul]
* MAST: Temp workaround in _updown.mast for martians problem [Bart/Paul]
* Cleanup source code to use C-style comments [Tuomo]
* Bugtracker bugs fixed:
   # 1120: [PATCH] netlink receive buffer size too small for linux 2.6.32
           [Roman Hoog Antink]

v2.6.27 (June 21, 2010)
* Fixes to the SAREF / MAST code to avoid recursion loops [Bart]
* KLIPS compile fixes for 2.4 kernels [David]

programs/showhostkey/showhostkey.c
/*
 * show the host keys in various formats
 * Copyright (C) 2005 Michael Richardson <mcr@xelerance.com>
 * Copyright (C) 1999, 2000, 2001  Henry Spencer.
 * Copyright (C) 2003-2008 Michael C Richardson <mcr@xelerance.com> 
 * Copyright (C) 2003-2010 Paul Wouters <paul@xelerance.com> 
 * Copyright (C) 2009 Avesh Agarwal <avagarwa@redhat.com>
 * Copyright (C) 2009 Stefan Arentz <stefan@arentz.ca>
 * Copyright (C) 2010 Tuomo Soini <tis@foobar.fi>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.  See <http://www.fsf.org/copyleft/gpl.txt>.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * for more details.
 *
 * replaces a shell script.
 *
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <limits.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <openswan.h>
#include <sys/utsname.h>

#include <arpa/nameser.h>
/* older versions lack ipseckey support */
#ifndef ns_t_ipseckey
# define ns_t_ipseckey 45
#endif

#include "constants.h"
#include "oswalloc.h"
#include "oswcrypto.h"
#include "oswlog.h"
#include "oswconf.h"
#include "secrets.h"
#include "mpzfuncs.h"

#ifdef HAVE_LIBNSS
# include <nss.h>
# include <prerror.h>
# include <prinit.h>
#endif

char usage[] = "Usage: ipsec showhostkey [--ipseckey {gateway}] [--left ] [--right ]\n"
             "                         [--dump ] [--list ] [--x509self]\n"
             "                         [--x509req ] [--x509cert ]      \n"
             "                         [ --txt gateway ] [--dhclient ] \n"
             "                         [ --file secretfile ] \n"
             "                         [ --keynum count ] [ --id identity ]\n"
             "                         [ --rsaid keyid ] [--verbose] [--version]\n";

struct option opts[] = {
  {"help",	no_argument,	NULL,	'?',},
  {"left",	no_argument,	NULL,	'l',},
  {"right",	no_argument,	NULL,	'r',},
  {"dump",	no_argument,	NULL,	'D',},
  {"list",	no_argument,	NULL,	'L',},
  {"x509self",	no_argument,	NULL,	's',},
  {"x509req",	no_argument,	NULL,	'R',},
  {"x509cert",	no_argument,	NULL,	'c',},
  {"txt",	required_argument,NULL,	't',},
  {"ipseckey",	required_argument,NULL,	'K',},
  {"dhclient",	no_argument,    NULL,	'd',},
  {"file",	required_argument,NULL,	'f',},
  {"keynum",	required_argument,NULL,	'n',},
  {"id",	required_argument,NULL,	'i',},
  {"rsaid",	required_argument,NULL,	'I',},
  {"version",	no_argument,	 NULL,	'V',},
  {"verbose",	no_argument,	 NULL,	'v',},
  {0,		0,	NULL,	0,}
};

char *progname = "ipsec showhostkey";	/* for messages */

void exit_tool(int code)
{
    tool_close_log();
    exit(code);
}

void
showhostkey_log(int mess_no, const char *message, ...)
{
    va_list args;

    va_start(args, message);
    vfprintf(stderr, message, args);
    va_end(args);
}


int print_key(struct secret *secret
	      , struct private_key_stuff *pks
	      , void *uservoid, bool disclose)
{
    int lineno = osw_get_secretlineno(secret);
    struct id_list *l = osw_get_idlist(secret);
    char idb[IDTOA_BUF];
    int count=1;

    char pskbuf[128];

    if(pks->kind == PPK_PSK || pks->kind==PPK_XAUTH) {
	datatot(pks->u.preshared_secret.ptr,
		pks->u.preshared_secret.len,
		'x', pskbuf, sizeof(pskbuf));
    }

    while(l) {
	idtoa(&l->id, idb, IDTOA_BUF);

	switch(pks->kind) {
	case PPK_PSK:
	    printf("%d(%d): PSK keyid: %s\n", lineno, count, idb);
	    if(disclose) printf("    psk: \"%s\"\n", pskbuf);
	    break;
	    
	case PPK_RSA:
	    printf("%d(%d): RSA keyid: %s with id: %s\n", lineno, count, pks->u.RSA_private_key.pub.keyid,idb);
	    break;
	    
	case PPK_XAUTH:
	    printf("%d(%d): XAUTH keyid: %s\n", lineno, count, idb);
	    if(disclose) printf("    xauth: \"%s\"\n", pskbuf);
	    break;
	    
	case PPK_PIN:
	    printf("%d:(%d) PIN key-type not yet supported for id: %s\n", lineno, count, idb);
	    break;
	}
	
	l=l->next;
	count++;
    }
    
    return 1;
}

int list_key(struct secret *secret,
	     struct private_key_stuff *pks,
	     void *uservoid)
{
    return print_key(secret, pks, uservoid, FALSE);
}

int dump_key(struct secret *secret,
	     struct private_key_stuff *pks,
	     void *uservoid)
{
    return print_key(secret, pks, uservoid, TRUE);
}


int pickbyid(struct secret *secret,
	     struct private_key_stuff *pks,
	     void *uservoid)
{
    char *rsakeyid = (char *)uservoid;

    if(strcmp(pks->u.RSA_private_key.pub.keyid, rsakeyid)==0) {
	return 0;
    }
    return 1;
}

struct secret *get_key_byid(struct secret *host_secrets, char *rsakeyid)
{
    return osw_foreach_secret(host_secrets, pickbyid, rsakeyid);
}

void list_keys(struct secret *host_secrets)
{
    (void)osw_foreach_secret(host_secrets, list_key, NULL);
}

char *get_default_keyid(struct secret *host_secrets)
{
    struct private_key_stuff *pks = osw_get_pks(host_secrets);

    return pks->u.RSA_private_key.pub.keyid;
}

     
void dump_keys(struct secret *host_secrets)
{
    (void)osw_foreach_secret(host_secrets, dump_key, NULL);
}

struct secret *pick_key(struct secret *host_secrets
			, char *idname)
{
    struct id id;
    struct secret *s;
    err_t e;

    e = atoid(idname, &id, FALSE);
    if(e) {
	printf("%s: key '%s' is invalid\n", progname, idname);
	exit(4);
    }

    s = osw_find_secret_by_id(host_secrets, PPK_RSA
			      , &id, NULL, TRUE /* asymmetric */);
    
    if(s==NULL) {
	char abuf[IDTOA_BUF];
	idtoa(&id, abuf, IDTOA_BUF);
	printf("%s: can not find key: %s (%s)\n", progname, idname, abuf);
	exit(5);
    }

    return s;
}

unsigned char *pubkey_to_rfc3110(const struct RSA_public_key *pub,
				 unsigned int *keybuflen)
{
    unsigned char *buf;
    unsigned char *p;
    unsigned int elen;

    chunk_t e, n;


    e = mpz_to_n2(&pub->e);
    n = mpz_to_n2(&pub->n);
    elen = e.len;

    buf = alloc_bytes(e.len+n.len+3, "buffer for rfc3110");
    p = buf;
    
    if (elen <= 255)
	*p++ = elen;
    else if ((elen &~ 0xffff) == 0) {
	*p++ = 0;
	*p++ = (elen>>8) & 0xff;
	*p++ = elen & 0xff;
    } else
	return 0;	/* unrepresentable exponent length */

    memcpy(p, e.ptr, e.len);
    p+=e.len;
    memcpy(p, n.ptr, n.len);
    p+=n.len;

    *keybuflen=(p-buf);

    return buf;
}    

void show_dnskey(struct secret *s
		 , char *idname
		 , int precedence
		 , char *gateway
		 , int rr_type)
{
    char qname[256];
    char base64[8192];
    const struct private_key_stuff *pks = osw_get_pks(s);
    unsigned char *keyblob;
    unsigned int keybloblen;

    gethostname(qname, sizeof(qname));

    if(pks->kind != PPK_RSA) {
	printf("%s: wrong kind of key %s in show_dnskey. Expected PPK_RSA.\n",
		progname, enum_name(&ppk_names, pks->kind));
	exit(5);
    }

    keyblob = pubkey_to_rfc3110(&pks->u.RSA_private_key.pub, &keybloblen);

    datatot(keyblob, keybloblen, 's', base64, sizeof(base64));

    switch(rr_type) {
    case ns_t_key:
	printf("key record type has been obsoleted\n");
	break;

    case ns_t_ipseckey:
	printf("ipseckey record type not yet implemented\n");
	break;

    case ns_t_txt:
	printf("; info about key: %s\n", 
	       pks->u.RSA_private_key.pub.keyid);
	printf("%s.    IN    TXT    \"X-IPsec-Server(%d)=%s \" ",
	       qname, precedence, gateway);
	{
	    int len = strlen(base64);
	    char *p=base64+2;  /* skip 0s */
	    while(len > 0) {
		int printlen = len;
		char saveit;
		if(printlen > 240) {
		    printlen=240;
		}
		saveit=p[printlen];
		p[printlen]='\0';
		printf("\"%s\" ",p);
		p[printlen]=saveit;
		p+=printlen;
		len-=printlen;
	    }
	}
	printf("\n");
	break;
    }
}
     
void show_confkey(struct secret *s
		  , char *idname
		  , char *side)
{
    char base64[8192];
    const struct private_key_stuff *pks = osw_get_pks(s);
    unsigned char *keyblob;
    unsigned int keybloblen;

    if(pks->kind != PPK_RSA) {
        char *enumstr = "gcc is crazy";
        switch (pks->kind) {
        case PPK_PSK:
                enumstr = "PPK_PSK";
        case PPK_PIN:
                enumstr = "PPK_PIN";
        case PPK_XAUTH:
                enumstr = "PPK_XAUTH";
        default:
                sscanf(enumstr,"UNKNOWN (%d)",(int *)pks->kind);
        }
	printf("%s: wrong kind of key %s in show_confkey. Expected PPK_RSA.\n", progname,enumstr);
	exit(5);
    }

    keyblob = pubkey_to_rfc3110(&pks->u.RSA_private_key.pub, &keybloblen);

    datatot(keyblob, keybloblen, 's', base64, sizeof(base64));

    printf("\t# rsakey %s\n", 
	   pks->u.RSA_private_key.pub.keyid);
    printf("\t%srsasigkey=%s\n", side,
	   base64);
}
     


int main(int argc, char *argv[])
{
    char secrets_file[PATH_MAX];
    int opt;
    int errflg = 0;
    bool left_flg=FALSE;
    bool right_flg=FALSE;
    bool dump_flg=FALSE;
    bool list_flg=FALSE;
    bool x509self_flg=FALSE;
    bool x509req_flg=FALSE;
    bool x509cert_flg=FALSE;
    bool txt_flg=FALSE;
    bool ipseckey_flg=FALSE;
    bool dhclient_flg=FALSE;
    char *gateway = NULL;
    int precedence=10;
    int verbose=0;
    const struct osw_conf_options *oco = osw_init_options();
    char *rsakeyid, *keyid;
    struct secret *host_secrets = NULL;
    struct secret *s;
    prompt_pass_t pass;

    rsakeyid=NULL;
    keyid=NULL;

    /* start up logging system */
    tool_init_log();

    snprintf(secrets_file, PATH_MAX, "%s/ipsec.secrets", oco->confdir);
    
    while ((opt = getopt_long(argc, argv, "", opts, NULL)) != EOF) {
	switch (opt) {
	case '?':
	    goto usage;
	    break;

	case 'l':
	    left_flg=TRUE;
	    break;
	case 'r':
	    right_flg=TRUE;
	    break;

	case 'D': /* --dump */
	    dump_flg=TRUE;
	    break;

	case 'K':
	    ipseckey_flg=TRUE;
	    gateway=clone_str(optarg, "gateway");
	    break;

	case 'L':
	    list_flg=TRUE;
	    break;

	case 's':
	case 'R':
	case 'c':
	    break;

	case 't':
	    txt_flg=TRUE;
	    gateway=clone_str(optarg, "gateway");
	    break;

	case 'd':
	    break;

	case 'f':  /* --file arg */
	    secrets_file[0]='\0';
	    strncat(secrets_file, optarg, PATH_MAX-1);
	    break;

	case 'i':
	    keyid=clone_str(optarg, "keyname");
	    break;

	case 'I':
	    rsakeyid=clone_str(optarg, "rsakeyid");
	    break;

	case 'n':
	case 'h':
	    break;

	case 'v':
	    verbose++;
	    break;

	case 'V':
	    fprintf(stdout, "%s\n", ipsec_version_string());
	    exit(0);

	default:
	    goto usage;
	}
    }

    if(errflg) {
    usage:
	fputs(usage, stderr);
	exit(1);
    }
    
    if(!left_flg && !right_flg && !dump_flg && !list_flg
       && !x509self_flg && !x509req_flg && !x509cert_flg && !txt_flg
       && !ipseckey_flg && !dhclient_flg) {
	fprintf(stderr, "You must specify some operation\n");
	goto usage;
    }
    
    if((left_flg + right_flg + dump_flg + list_flg
	+ x509self_flg + x509req_flg + x509cert_flg + txt_flg
	+ ipseckey_flg + dhclient_flg) > 1) {
	fprintf(stderr, "You must specify only one operation\n");
	goto usage;
    }

    if(verbose > 2) {
#ifdef DEBUG
	set_debugging(DBG_ALL);
#else
	fprintf(stderr, "verbosity cannot be set - compiled without DEBUG\n");
#endif
    }

    /* now load file from indicated location */
    pass.prompt=showhostkey_log;
    pass.fd = 2; /* stderr */

#ifdef HAVE_LIBNSS
    PRBool nss_initialized = PR_FALSE;
    SECStatus rv;
    char buf[100];
    snprintf(buf, sizeof(buf), "%s",oco->confddir);
    loglog(RC_LOG_SERIOUS,"nss directory showhostkey: %s",buf);
    PR_Init(PR_USER_THREAD, PR_PRIORITY_NORMAL, 1);
    if ((rv = NSS_InitReadWrite(buf)) != SECSuccess) {
	fprintf(stderr, "%s: NSS_InitReadWrite returned %d\n",progname, PR_GetError());
	exit(1);
    }
   nss_initialized = PR_TRUE; 
   PK11_SetPasswordFunc(getNSSPassword); 
#endif

    load_oswcrypto();
    osw_load_preshared_secrets(&host_secrets, verbose>0?TRUE:FALSE,
			       secrets_file, &pass);

#ifdef HAVE_LIBNSS
    if (nss_initialized) {
	NSS_Shutdown();
    }
    PR_Cleanup();
#endif

    /* options that apply to entire files */
    if(dump_flg) {
	/* dumps private key info too */
	dump_keys(host_secrets);
	exit(0);
    }

    if(list_flg) {
	list_keys(host_secrets);
	exit(0);
    }

    if(rsakeyid) {
	printf("; picking by rsakeyid=%s\n", rsakeyid);
	s = get_key_byid(host_secrets, rsakeyid);
	keyid=rsakeyid;
    } else if(keyid) {
	printf("; picking by keyid=%s\n", keyid);
	s = pick_key(host_secrets, keyid);
    } else {
	/* Paul: This assumption is WRONG. Mostly I have PSK's above my
 	 * multiline default : RSA entry, and then this assumption breaks
 	 * The proper test would be for ": RSA" vs "@something :RSA"
 	 */
	/* default key is the *LAST* key, because it is first in the file.*/
	s=osw_get_defaultsecret(host_secrets);
	keyid="default";
    }

    if(s==NULL) {
	printf("No keys found\n");
	exit(20);
    }

    if(left_flg) {
	show_confkey(s, keyid, "left");
	exit(0);
    }

    if(right_flg) {
	show_confkey(s, keyid, "right");
	exit(0);
    }

    if(ipseckey_flg || txt_flg) {
	int rr_type = ns_t_invalid;
	if(ipseckey_flg) {
	    rr_type = ns_t_ipseckey;
	} else if(txt_flg) {
	    rr_type = ns_t_txt;
	}
	show_dnskey(s, keyid,
		    precedence, gateway,
		    rr_type);
    }

    exit(0);
}

/*
 * Local Variables:
 * c-basic-offset:4
 * c-style: pluto
 * End:
 */
